_posts/2025-12-12-iceberg-in-the-browser.md

```diff
@@ -0,0 +1,119 @@
+
+![Iceberg with DuckDB](/images/blog/iceberg-wasm/iceberg-with-duckdb-dark.svg){: .darkmode-img }
+![Iceberg with DuckDB](/images/blog/iceberg-wasm/iceberg-with-duckdb-light.svg){: .lightmode-img }
+
+DuckDB supports both Iceberg interaction models.
+In the _client–server model,_ DuckDB runs on the server to read the Iceberg datasets.
+From the user's point of view, the choice of engine is transparent, and DuckDB is just one of many engines that the server could use in the background.
+The *client-is-the-server* model is more interesting:
+here, users [install a DuckDB client locally]({% link install/index.html %})
+and use it through its SQL interface to query Iceberg catalogs.
+For example:
+
+```sql
+CREATE SECRET test_secret (
+    TYPE S3, 
+    KEY_ID '⟨AKIAIOSFODNN7EXAMPLE⟩',
+    SECRET '⟨wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY⟩'
+);
+
+ATTACH '⟨warehouse⟩' AS db (
+    TYPE iceberg,
+    ENDPOINT_URL '⟨https://your-iceberg-endpoint⟩',
+);
+
+SELECT sum(value)
+FROM db.table
+WHERE other_column = '⟨some_value⟩';
+```
+
+> You can discover the full DuckDB-Iceberg extension feature set, including insert and update capabilities, in our [earlier blog post]({% post_url 2025-11-28-iceberg-writes-in-duckdb %}).
+
+## Iceberg with DuckDB in the Browser
+
+While setting up a local DuckDB installation is quite simple, opening a browser tab is even quicker.
+Therefore, we asked ourselves: could we support the *client-is-the-server* model directly from within a browser tab?
+This could provide zero-setup, no-infrastructure, properly serverless option for interacting with Iceberg catalogs.
+
+![Iceberg with DuckDB-Wasm](/images/blog/iceberg-wasm/duckdb-iceberg-with-duckdb-wasm-dark.svg){: .darkmode-img }
+![Iceberg with DuckDB-Wasm](/images/blog/iceberg-wasm/duckdb-iceberg-with-duckdb-wasm-light.svg){: .lightmode-img }
+
+Luckily, DuckDB has a client that can run in any browser!
+[DuckDB-Wasm]({% link docs/stable/clients/wasm/overview.md %}) is a WebAssembly port of DuckDB, which [supports loading of extensions]({% post_url 2023-12-18-duckdb-extensions-in-wasm %}).
+
+Interacting with a Iceberg REST Catalog requires a number of functionalities. First, the ability to talk to a REST API over HTTP(S). Second, the ability to read and write `avro` and `parquet` files on object storage. Third, negotiating authentication to access those resources on behalf of the user. All of these can be done also from a browser without the need to call a native component.
+
+To support these, we implemented the following high-level changes:
+
+* In the core `duckdb` codebase, we redesigned HTTP interactions, so that extensions and clients have a uniform interface to the networking stack.
+* In `duckdb-wasm`, we implemented such an interface, which in this case is a wrapper around the available JavaScript network stack.
+* In `duckdb-iceberg`, we routed all networking through the common HTTP interface, so that both in native and Wasm the same logic is executed.
```

Comment:
```suggestion
* In `duckdb-iceberg`, we routed all networking through the common HTTP interface, so that native DuckDB and DuckDB-wasm execute the same logic.
```